### PR TITLE
Added support for data stream monitoring for kafkajs

### PR DESCRIPTION
**Pull request by ida613**
### What does this PR do?
Modifies kafkajs plugin edge tags based off of dsm system tests

### Remaining TODOs

- [x] probably keep gzip
- [x] figure out messagepack - if we keep it it's 183kb zipped - probably keep for now
- [x] downgrade mocha (easy)
- [x] fix two rebase issues (easy)
- [x] pull DSM code from Tracer (medium)
- [x] makeRequest remove copypasta (medium) - maybe, doesn't seem that beneficial
- [x] switch to a different hashing algo (easy)

### CODE REVIEW

File: packages/dd-trace/src/telemetry/index.js [hunk 1]
```diff
@@ -52,8 +52,6 @@ function flatten (input, result = [], prefix = [], traversedObjects = null) {
 
 function appStarted () {
   return {
-    integrations: getIntegrations(),
```
Comment (tlhunter) [inline packages/dd-trace/src/telemetry/index.js:55]: TODO: Accidental commit

File: package.json [hunk 2]
```diff
@@ -89,8 +91,10 @@
     "lru-cache": "^7.14.0",
     "methods": "^1.1.2",
     "module-details-from-path": "^1.0.3",
+    "msgpack-lite": "^0.1.26",
```
Comment (rochdev) [inline package.json:94]: We shouldn't be using this outside of devDependencies as MessagePack is encoded manually.

File: package.json [hunk 3]
```diff
@@ -89,8 +91,10 @@
...
     "node-abort-controller": "^3.0.1",
     "opentracing": ">=0.12.1",
+    "pako": "^2.1.0",
```
Comment (rochdev) [inline package.json:97]: Why is this needed?

File: package.json [hunk 4]
```diff
@@ -76,8 +76,10 @@
     "@opentelemetry/core": "<1.4.0",
     "crypto-randomuuid": "^1.0.0",
     "diagnostics_channel": "^1.1.0",
+    "fnv-plus": "^1.3.1",
```
Comment (rochdev) [inline package.json:79]: Do we need this? If we can use any hashing algorithm, why not use the same as what AppSec is using?

File: package.json [hunk 5]
```diff
@@ -120,12 +124,11 @@
     "get-port": "^3.2.0",
     "glob": "^7.1.6",
     "graphql": "0.13.2",
-    "int64-buffer": "^0.1.9",
+    "gzip-js": "^0.3.2",
```
Comment (rochdev) [inline package.json:127]: Similar to pako, why is this needed?

File: package.json [hunk 6]
```diff
@@ -120,12 +124,11 @@
...
     "jszip": "^3.5.0",
     "knex": "^2.4.2",
     "mkdirp": "^0.5.1",
-    "mocha": "8",
-    "msgpack-lite": "^0.1.26",
+    "mocha": "^10.2.0",
```
Comment (rochdev) [inline package.json:131]: This breaks compatibility with Node 12.

File: packages/datadog-plugin-kafkajs/src/consumer.js [hunk 7]
```diff
@@ -6,10 +6,13 @@ class KafkajsConsumerPlugin extends ConsumerPlugin {
...
 
-  start ({ topic, partition, message }) {
+  start ({ topic, partition, message, groupId }) {
     const childOf = extract(this.tracer, message.headers)
-    this.startSpan({
+    const service = this.tracer._service
+
+    const header = {
       childOf,
+      service: this.config.service || `${service}-kafka`,
```
Comment (rochdev) [inline packages/datadog-plugin-kafkajs/src/consumer.js:15]: Why is this being changed in this PR?

File: packages/datadog-plugin-kafkajs/src/consumer.js [hunk 8]
```diff
@@ -20,10 +23,18 @@ class KafkajsConsumerPlugin extends ConsumerPlugin {
       metrics: {
         'kafka.partition': partition
       }
-    })
+    }
+
+    if (this.config.dsmEnabled) {
+      this.tracer.decodeDataStreamsContext(message.headers['dd-pathway-ctx'])
```
Comment (rochdev) [inline packages/datadog-plugin-kafkajs/src/consumer.js:29]: Why is this exposed by the tracer? Is this a public API? Also, the current implementation is Kafka-specific, so is this really a generic API?

File: packages/datadog-plugin-kafkajs/src/consumer.js [hunk 9]
```diff
@@ -20,10 +23,18 @@ class KafkajsConsumerPlugin extends ConsumerPlugin {
...
+      this.tracer
+        .setCheckpoint(['direction:in', `group:${groupId}`, `topic:${topic}`, 'type:kafka'])
+    }
+    this.startSpan('kafka.consume', header)
   }
 }
 
+// split this into two
```
Comment (rochdev) [inline packages/datadog-plugin-kafkajs/src/consumer.js:37]: Is this a TODO? If it is it should be prefixed as such and with more context.

File: packages/datadog-plugin-kafkajs/src/producer.js [hunk 10]
```diff
@@ -1,13 +1,22 @@
...
 
 class KafkajsProducerPlugin extends ProducerPlugin {
   static get id () { return 'kafkajs' }
   static get operation () { return 'produce' }
 
   start ({ topic, messages }) {
-    const span = this.startSpan({
+
+    let pathwayCtx
+    if (this.config.dsmEnabled) {
```
Comment (rochdev) [inline packages/datadog-plugin-kafkajs/src/producer.js:13]: This block could be extracted to the base class as a helper.

File: packages/datadog-plugin-kafkajs/src/producer.js [hunk 11]
```diff
@@ -1,13 +1,22 @@
...
+      const dataStreamsContext = this.tracer
+        .setCheckpoint(['direction:out', `topic:${topic}`, 'type:kafka'])
+      pathwayCtx = encodePathwayContext(dataStreamsContext)
+    }
+    const span = this.startSpan('kafka.produce', {
+      service: this.config.service || `${this.tracer._service}-kafka`,
```
Comment (rochdev) [inline packages/datadog-plugin-kafkajs/src/producer.js:19]: Similar to the above, why are we changing the service name logic?

File: packages/datadog-plugin-kafkajs/src/producer.js [hunk 12]
```diff
@@ -20,6 +29,7 @@ class KafkajsProducerPlugin extends ProducerPlugin {
 
     for (const message of messages) {
       if (typeof message === 'object') {
+        if (this.config.dsmEnabled) message.headers['dd-pathway-ctx'] = pathwayCtx
```
Comment (rochdev) [inline packages/datadog-plugin-kafkajs/src/producer.js:32]: Similar to the above, this could be hidden away in a base class method.

File: packages/datadog-plugin-kafkajs/test/hash.spec.js [hunk 13]
```diff
@@ -0,0 +1,50 @@
+'use strict'
```
Comment (rochdev) [inline packages/datadog-plugin-kafkajs/test/hash.spec.js:1]: Can this be tested using the plugin instead?

File: packages/dd-trace/src/datastreams/pathway.js [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+}
+
+function computeHash (service, env, edgeTags, parentHash) {
+  const currentHash = fnvHash(`${service}${env}` + edgeTags.join(''))
+  const buf = Buffer.concat([ currentHash, parentHash ], 16)
+  return fnvHash(buf.toString())
+}
+
+function encodePathwayContext (dataStreamsContext) {
+  return Buffer.concat([ dataStreamsContext.hash, Buffer.from(encodeVarint(Math.round(dataStreamsContext.pathwayStartNs / 1e6))), Buffer.from(encodeVarint(Math.round(dataStreamsContext.edgeStartNs / 1e6))) ], 20)
```
Comment (rochdev) [inline packages/dd-trace/src/datastreams/pathway.js:19]: This would be more readable if split in multiple lines.

File: packages/dd-trace/src/datastreams/pathway.js [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+// this inconsistency is ok because hashes do not need to be consistent across services
+const fnv = require('fnv-plus')
+const { encodeVarint, decodeVarint } = require('./encoding')
+
+function fnvHash (checkpointString) {
+  const hash = fnv.hash(checkpointString, 64)
+  return Buffer.from(hash.hex(), 'hex')
+}
+
+function computeHash (service, env, edgeTags, parentHash) {
```
Comment (rochdev) [inline packages/dd-trace/src/datastreams/pathway.js:12]: Could this be cached? Otherwise what if you are sending thousands of messages to Kafka? Should we also add a benchmark to see how much overhead this is adding?

File: packages/dd-trace/src/datastreams/processor.js [hunk 16]
```diff
@@ -0,0 +1,193 @@
...
+      port,
+      url
+    })
+    this.bucketSizeNs = 1e10
+    this.buckets = new TimeBuckets()
+    this.hostname = os.hostname()
+    this.enabled = dsmEnabled
+    this.env = env
+    this.tags = tags || {}
+    this.service = this.tags.service || 'unnamed-nodejs-service'
```
Comment (rochdev) [inline packages/dd-trace/src/datastreams/processor.js:82]: Why is the default service name repeated here?

File: packages/dd-trace/src/datastreams/processor.js [hunk 17]
```diff
@@ -0,0 +1,193 @@
...
+      Lang: 'javascript'
+    }
+    this.writer.flush(payload)
+  }
+
+  recordCheckpoint (checkpoint) {
+    if (!this.enabled) return
+    const bucketTime = Math.round(checkpoint.currentTimestamp - (checkpoint.currentTimestamp % this.bucketSizeNs))
+    this.buckets.forTime(bucketTime)
+      .forCheckpoint(checkpoint)
```
Comment (rochdev) [inline packages/dd-trace/src/datastreams/processor.js:108]: Is this actually by checkpoint? This seems fairly expensive, unless we don't expect a lot of checkpoints?

File: packages/dd-trace/src/datastreams/writer.js [hunk 18]
```diff
@@ -0,0 +1,66 @@
...
+  }
+
+  flush (payload) {
+    if (!request.writable) {
+      log.debug(() => `Maximum number of active requests reached. Payload discarded: ${JSON.stringify(payload)}`)
+      return
+    }
+    const encodedPayload = msgpack.encode(payload, { codec })
+
+    zlib.gzip(encodedPayload, { level: 1 }, (err, compressedData) => {
```
Comment (rochdev) [inline packages/dd-trace/src/datastreams/writer.js:49]: Why is this gzipped? Historically, we've avoided gzipping to the agent because the CPU overhead is not worth the bandwidth reduction on a local network.

File: packages/dd-trace/src/datastreams/writer.js [hunk 19]
```diff
@@ -0,0 +1,66 @@
...
+      if (err) {
+        log.error(err)
+        return
+      }
+      makeRequest(compressedData, this._url, (err, res) => {
+        log.debug(`Response from the intake: ${res}`)
```
Comment (rochdev) [inline packages/dd-trace/src/datastreams/writer.js:55]: The request is not sent to the intake, I'm not sure I understand this error.

File: packages/dd-trace/src/datastreams/writer.js [hunk 20]
```diff
@@ -0,0 +1,66 @@
+const pkg = require('../../../../package.json')
+const log = require('../log')
+const request = require('../exporters/common/request')
+const { URL, format } = require('url')
+const msgpack = require('msgpack-lite')
+const zlib = require('zlib')
+const codec = msgpack.createCodec({ int64: true })
+
+function makeRequest (data, url, cb) {
```
Comment (rochdev) [inline packages/dd-trace/src/datastreams/writer.js:9]: Why is an entire new `makeRequest` required? Why can't the existing code be reused?

File: packages/dd-trace/src/scope.js [hunk 21]
```diff
@@ -34,6 +34,15 @@ class Scope {
     }
   }
 
+  getDataStreamsContext () {
```
Comment (rochdev) [inline packages/dd-trace/src/scope.js:37]: These should not go on the scope manager as it's meant only for span management.

File: packages/dd-trace/src/tracer.js [hunk 22]
```diff
@@ -17,11 +19,24 @@ const MEASURED = tags.MEASURED
 class DatadogTracer extends Tracer {
   constructor (config) {
     super(config)
-
+    this._dataStreamsProcessor = new DataStreamsProcessor(config)
     this._scope = new Scope()
     setStartupLogConfig(config)
   }
 
+  setCheckpoint (edgeTags) {
```
Comment (rochdev) [inline packages/dd-trace/src/tracer.js:27]: Similar to the checkpoint, this should not be on the tracer.

**Review by rochdev — CHANGES_REQUESTED** [review]

File: packages/datadog-plugin-kafkajs/src/consumer.js (hunk 7 above)
Comment (tlhunter) [reply packages/datadog-plugin-kafkajs/src/consumer.js:15]: This is probably a rebase error

File: packages/datadog-plugin-kafkajs/src/producer.js (hunk 11 above)
Comment (tlhunter) [reply packages/datadog-plugin-kafkajs/src/producer.js:19]: Another rebase issue

File: packages/dd-trace/src/datastreams/writer.js (hunk 18 above)
Comment (tlhunter) [reply packages/dd-trace/src/datastreams/writer.js:49]: @piochelepiotr should have insight to the protocol

File: package.json (hunk 4 above)
Comment (ida613) [reply package.json:79]: which one is AppSec using?

File: package.json (hunk 2 above)
Comment (tlhunter) [reply package.json:94]: @piochelepiotr added this, I suspect he didn't know we already had a message pack implementation. I don't know if there are any shortcomings to our implementation that needed to be solved with a third party package. Piotr is out this week so we'll try to investigate.

File: package.json (hunk 3 above)
Comment (tlhunter) [reply package.json:97]: This isn't needed anymore, it's leftover from some refactor

File: package.json (hunk 4 above)
Comment (tlhunter) [reply package.json:79]: Specifically, fnv-plus implements the FNV64 1a hash while the "spec" basically says we should use FNV64 1 otherwise we must use some sort of consistent hash. Since FNV64 1a is no better than any other hash, say CRC32 or SHA1, then we might as well just use an existing hash and reduce our dependency count.

So, we'll figure out what hash appsec already introduced and switch to that.

File: package.json (hunk 5 above)
Comment (tlhunter) [reply package.json:127]: This is leftover from some refactor and will be removed

File: package.json (hunk 6 above)
Comment (tlhunter) [reply package.json:131]: After looking through the code we didn't notice anything that requires a new version of Mocha. There aren't any notes as to why Mocha was upgraded. For that reason we'll try downgrading the version and checking if the tests still run.

File: packages/datadog-plugin-kafkajs/src/consumer.js (hunk 8 above)
Comment (tlhunter) [reply packages/datadog-plugin-kafkajs/src/consumer.js:29]: I don't think that this is intended to be public. We will refactor it to make it no longer a method on the tracer object.

File: packages/datadog-plugin-kafkajs/src/producer.js (hunk 12 above)
Comment (tlhunter) [reply packages/datadog-plugin-kafkajs/src/producer.js:32]: I think we'll move it once we start supporting more queue products next quarter

File: packages/datadog-plugin-kafkajs/src/producer.js (hunk 10 above)
Comment (tlhunter) [reply packages/datadog-plugin-kafkajs/src/producer.js:13]: I think we'll do this when we support more queues

File: packages/datadog-plugin-kafkajs/test/hash.spec.js (hunk 13 above)
Comment (tlhunter) [reply packages/datadog-plugin-kafkajs/test/hash.spec.js:1]: I think you're saying that we're testing this functionality at a lower level, specifically the hash functionality itself, instead of at a higher level, being the Kafka/DMS code that calls the hashes.

We could go through the effort to refactor all these tests to move where the testing happens. But before we proceed, what would the value gained be?

File: packages/dd-trace/src/datastreams/pathway.js (hunk 14 above)
Comment (tlhunter) [reply packages/dd-trace/src/datastreams/pathway.js:19]:
```js
  return Buffer.concat([
    dataStreamsContext.hash,
    Buffer.from(encodeVarint(Math.round(dataStreamsContext.pathwayStartNs / 1e6))),
    Buffer.from(encodeVarint(Math.round(dataStreamsContext.edgeStartNs / 1e6)))
  ], 20)
```

File: packages/dd-trace/src/datastreams/pathway.js (hunk 15 above)
Comment (tlhunter) [reply packages/dd-trace/src/datastreams/pathway.js:12]: We can use lru-cache to cache these

```js
function computeHash (service, env, edgeTags, parentHash) {
  const key = `${service}${env}` + edgeTags.join('') + parentHash.toString

  if (cache.has(key)) {
    return cache.get(key)
  }
  const currentHash = fnvHash(`${service}${env}` + edgeTags.join(''))
  const buf = Buffer.concat([ currentHash, parentHash ], 16)
  const val = fnvHash(buf.toString())
  cache.set(key, val)
  return val
}
```
```

File: packages/dd-trace/src/datastreams/processor.js (hunk 17 above)
Comment (tlhunter) [reply packages/dd-trace/src/datastreams/processor.js:108]: Are you asking if this is run once for every checkpoint in a chain of services? Like `svc1 -> svc2 -> svc3`? If so, this isn't the case, and the code isn't iterating for each of the prior services.

File: package.json (hunk 4 above)
Comment (ida613) [reply package.json:79]: Hashing algorithm appsec is using is [here](https://github.com/DataDog/dd-trace-js/blob/e4b9a268f0429b6f1e92c384b61a0d104aeb1259/packages/dd-trace/src/appsec/iast/analyzers/vulnerability-analyzer.js#L113), I'm not sure if it's a good candidate, or maybe we should just use a built-in algo

File: packages/datadog-plugin-kafkajs/src/consumer.js (hunk 7 above)
Comment (ida613) [reply packages/datadog-plugin-kafkajs/src/consumer.js:15]: `service` was used for setting checkpoints but it doesn't need to be in the header. I'm still trying to see if it's safe to delete after Piotr's change, my current take is it is. But let's see.

File: packages/dd-trace/src/datastreams/processor.js (hunk 17 above)
Comment (ida613) [reply packages/dd-trace/src/datastreams/processor.js:108]: I think the function is better named as "forThisCheckpoint"

File: packages/datadog-plugin-kafkajs/src/consumer.js (hunk 7 above)
Comment (tlhunter) [reply packages/datadog-plugin-kafkajs/src/consumer.js:15]: @ida613 this might not have been an intentional add. It did come up when we did that rebase.

File: packages/dd-trace/src/telemetry/index.js (hunk 1 above)
Comment (ida613) [reply packages/dd-trace/src/telemetry/index.js:55]: need to add back integrations + dependencies

**Comment by tlhunter** [comment]
Since we don't care about using the exact same hashing algorithm, but we do want values to be consistent and generated quickly, we can make use of MD5, and then only take the first 8 bytes from that.

```js
const crypto = require('crypto')

function simpleHash (str) {
  return crypto
    .createHash('md5')
    .update(str)
    .digest()
    .slice(0, 8)
}
```

File: packages/dd-trace/src/datastreams/writer.js (hunk 20 above)
Comment (tlhunter) [reply packages/dd-trace/src/datastreams/writer.js:9]: We already have two versions and this new version is different enough from the existing ones that I don't think we benefit much from code reuse.

File: package.json (hunk 2 above)
Comment (ida613) [reply package.json:94]: we've decided to keep this package since writing our custom message pack implementation for dsm would take too much effort at this point

File: package.json (hunk 4 above)
Comment (ida613) [reply package.json:79]: We have decided to use a built-in algo, will update it ^ ^

File: package.json (hunk 6 above)
Comment (piochelepiotr) [reply package.json:129]: I just ran `npm install msgpack-lite --save`, and this dependency got upgraded.

File: packages/dd-trace/src/scope.js (hunk 21 above)
Comment (piochelepiotr) [reply packages/dd-trace/src/scope.js:37]: I extended `scope` to be the full context, not only span scope, but also data streams scope.
Where would you see that data streams context live instead?

File: packages/dd-trace/src/datastreams/writer.js (hunk 18 above)
Comment (piochelepiotr) [reply packages/dd-trace/src/datastreams/writer.js:49]: With the protobuf sketches, the encoded payload can have a lot of 0, so that's why we chose to add a gzip in the first implementation.
I don't think it should have a big perf impact because:
1. We're using level 1 encoding
2. The amount of data sent is much much lower than traces, because we aggregate before sending.

File: package.json (hunk 4 above)
Comment (piochelepiotr) [reply package.json:79]: We can keep the cryptographic hash for now since there is a cache around it. It's not a big issue if it's slow.

File: packages/dd-trace/src/scope.js (hunk 21 above)
Comment (piochelepiotr) [reply packages/dd-trace/src/scope.js:37]: Moved outside of scope.

File: packages/dd-trace/src/tracer.js (hunk 22 above)
Comment (piochelepiotr) [reply packages/dd-trace/src/tracer.js:27]: Talked about it on zoom, we will keep that for now in the tracer, but not expose it to avoid too many changes here.

File: packages/dd-trace/test/datastreams/processor.spec.js [hunk 23]
```diff
@@ -0,0 +1,177 @@
...
+const DEFAULT_LATENCY = 100000000
+const DEFAULT_PARENT_HASH = Buffer.from('e858292fd15a41e4', 'hex')
+const ANOTHER_PARENT_HASH = Buffer.from('e858292fd15a4100', 'hex')
+const DEFAULT_CURRENT_HASH = Buffer.from('e858212fd11a41e5', 'hex')
+const ANOTHER_CURRENT_HASH = Buffer.from('e851212fd11a21e9', 'hex')
+
+const writer = {
+  flush: sinon.stub()
+}
+const DataStreamsWriter = sinon.stub().returns(writer)
```
Comment (tlhunter) [inline packages/dd-trace/test/datastreams/processor.spec.js:23]: This stays permanently stubbed for the duration of the test suite. Can you move all of this into a `before` / `beforeEach` block and then add the `.restore` in the equivalent `after` / `afterEach` block?

Comment (tlhunter) [reply packages/dd-trace/test/datastreams/processor.spec.js:23]: Also, you may need to wrap all of these describes into a single overarching describe for the entire file, with the before/after inside of the top level describe.

Comment (tlhunter) [reply packages/dd-trace/test/datastreams/processor.spec.js:23]: Sorry, I'm wrong, there's nothing to restore here

File: packages/dd-trace/src/datastreams/writer.js (hunk 18 above)
Comment (rochdev) [reply packages/dd-trace/src/datastreams/writer.js:49]: If it's not a lot of data, then why even bother with gzip?

File: package.json (hunk 2 above)
Comment (rochdev) [reply package.json:93]: The work has already been done in https://github.com/DataDog/dd-trace-js/blob/master/packages/dd-trace/src/encode/span-stats.js

**Review by rochdev — APPROVED** [review]
